dll/win32/shell32/CFolder.cpp

```diff
@@ -65,35 +53,39 @@ HRESULT STDMETHODCALLTYPE CFolder::get_Title(BSTR *pbs)
 HRESULT STDMETHODCALLTYPE CFolder::get_Application(IDispatch **ppid)
 {
     TRACE("(%p, %p)\n", this, ppid);
-    return E_NOTIMPL;
+
+    if (!ppid)
+        return E_INVALIDARG;
+
+    *ppid = m_Application;
+    (*ppid)->AddRef();
+
+    return S_OK;
 }
 
 HRESULT STDMETHODCALLTYPE CFolder::get_Parent(IDispatch **ppid)
 {
     TRACE("(%p %p)\n", this, ppid);
+
+    if (ppid)
+        *ppid = NULL;
+
     return E_NOTIMPL;
 }
 
 HRESULT STDMETHODCALLTYPE CFolder::get_ParentFolder(Folder **ppsf)
 {
     TRACE("(%p, %p)\n", this);
+
+    *ppsf = NULL;
+
     return E_NOTIMPL;
 }
 
 HRESULT STDMETHODCALLTYPE CFolder::Items(FolderItems **ppid)
 {
-    CFolderItems* items = new CComObject<CFolderItems>();
-    items->AddRef();
-
-    HRESULT hr = items->Init(ILClone(m_idlist));
-    if (FAILED_UNEXPECTEDLY(hr))
-    {
-        items->Release();
-        return hr;
-    }
-
-    *ppid = items;
-    return S_OK;
+    /* FolderItems_Constructor */
+    return ShellObjectCreatorInit<CFolderItems>(ILClone(m_idlist), this, IID_PPV_ARG(FolderItems, ppid));
```

Comment:
I would argue it's more logical to do the ILClone in the Init function rather than its caller.